### PR TITLE
Refactor Message Chains

### LINKED ISSUE #1: changeState() Message Chain

### LINKED ISSUE #2: moveToNextMachine() message chain

### LINKED ISSUE #3: Reduce occurrences of machine[theMachine] in MachineShopSimulator.java by making it it's own variable

### LINKED ISSUE #4: Refactor Message Chains

### PR DESCRIPTION
**Pull request by nickplucker**
This pull request is part of the work required to clean up the message chains in MachineShopSimulator.java. To achieve this, we used Intellij's refactoring tool 'extract method' to increase the codes readability. 

Closes #1 Closes #2 Closes #3 Closes #4

### CODE REVIEW

File: src/applications/MachineShopSimulator.java [hunk 1]
```diff
@@ -28,9 +28,9 @@ static boolean moveToNextMachine(Job theJob, SimulationResults simulationResults
         if (theJob.getTaskQ().isEmpty()) {// no next task
             simulationResults.setJobCompletionData(theJob.getId(), timeNow, timeNow - theJob.getLength());
             return false;
-        } else {// theJob has a next task
-                // get machine for next task
-            int p = ((Task) theJob.getTaskQ().getFrontElement()).getMachine();
+        } else {
+            // theJob has a next task get machine for next task
+            int p = getMachine(theJob);
```
Comment (YujingSong) [inline src/applications/MachineShopSimulator.java]: I like the way you shorter the chain, it is much more easier to read now.

File: src/applications/MachineShopSimulator.java [hunk 2]
```diff
@@ -42,40 +42,55 @@ static boolean moveToNextMachine(Job theJob, SimulationResults simulationResults
...
+    private static void setReturnTime(Machine machine) {
+        machine.setTotalWait(machine.getTotalWait() + timeNow
+                - machine.getActiveJob().getArrivalTime());
+    }
+
+    private static void getNextJob(Machine machine) {
+        machine.setActiveJob((Job) machine.getJobQ()
+                .remove());
+    }
+
```
Comment (YujingSong) [inline src/applications/MachineShopSimulator.java:80]: I do like these help methods, they make each method only focus on one thing.

File: src/applications/MachineShopSimulator.java [hunk 3]
```diff
@@ -42,40 +42,55 @@ static boolean moveToNextMachine(Job theJob, SimulationResults simulationResults
...
-                machine[theMachine].setActiveJob((Job) machine[theMachine].getJobQ()
-                        .remove());
-                machine[theMachine].setTotalWait(machine[theMachine].getTotalWait() + timeNow
-                        - machine[theMachine].getActiveJob().getArrivalTime());
-                machine[theMachine].setNumTasks(machine[theMachine].getNumTasks() + 1);
-                int t = machine[theMachine].getActiveJob().removeNextTask();
+            else {
+                // take job off the queue and work on it
+                getNextJob(currentMachine);
+                setReturnTime(currentMachine);
```
Comment (isaacyoa) [inline src/applications/MachineShopSimulator.java]: This is a lot more clear and really resolves the chain.

**Review by isaacyoa — CHANGES_REQUESTED** [review]
The refactoring benefits the code quality and makes everything much more clear to read/understand. 
I would go back and make sure all of the changes in your files are specific for the smell. There are scraps in xml files that are not necessary for the smell you are addressing.

File: .idea/modules.xml [hunk 4]
```diff
@@ -5,6 +5,7 @@
       <module fileurl="file://$PROJECT_DIR$/.idea/Machine-shop-simulator.iml" filepath="$PROJECT_DIR$/.idea/Machine-shop-simulator.iml" />
       <module fileurl="file://$PROJECT_DIR$/.idea/modules/Machine-shop-simulator_main.iml" filepath="$PROJECT_DIR$/.idea/modules/Machine-shop-simulator_main.iml" group="Machine-shop-simulator" />
       <module fileurl="file://$PROJECT_DIR$/.idea/modules/Machine-shop-simulator_test.iml" filepath="$PROJECT_DIR$/.idea/modules/Machine-shop-simulator_test.iml" group="Machine-shop-simulator" />
+      <module fileurl="file://$PROJECT_DIR$/.idea/modules/smells-in-machine-shop-simulator-team-name.iml" filepath="$PROJECT_DIR$/.idea/modules/smells-in-machine-shop-simulator-team-name.iml" />
```
Comment (globbycodes) [inline .idea/modules.xml]: That change is probably not related to any of your issues

File: .idea/modules/Machine-shop-simulator_main.iml [hunk 5]
```diff
@@ -5,6 +5,7 @@
     <exclude-output />
     <content url="file://$MODULE_DIR$/../../src">
       <sourceFolder url="file://$MODULE_DIR$/../../src" isTestSource="false" />
+      <sourceFolder url="file://$MODULE_DIR$/../../src/main/resources" type="java-resource" />
```
Comment (globbycodes) [inline .idea/modules/Machine-shop-simulator_main.iml]: This is also not related to the issue

File: .idea/modules/Machine-shop-simulator_test.iml [hunk 6]
```diff
@@ -3,7 +3,9 @@
   <component name="NewModuleRootManager" LANGUAGE_LEVEL="JDK_1_8">
     <output-test url="file://$MODULE_DIR$/../../out/test/classes" />
     <exclude-output />
-    <content url="file://$MODULE_DIR$/../../src/test" />
+    <content url="file://$MODULE_DIR$/../../src/test">
+      <sourceFolder url="file://$MODULE_DIR$/../../src/test/resources" type="java-test-resource" />
```
Comment (globbycodes) [inline .idea/modules/Machine-shop-simulator_test.iml]: This also should not be included

File: .idea/modules/Machine-shop-simulator_test.iml [hunk 7]
```diff
@@ -3,7 +3,9 @@
   <component name="NewModuleRootManager" LANGUAGE_LEVEL="JDK_1_8">
     <output-test url="file://$MODULE_DIR$/../../out/test/classes" />
     <exclude-output />
-    <content url="file://$MODULE_DIR$/../../src/test" />
+    <content url="file://$MODULE_DIR$/../../src/test">
```
Comment (holma198) [inline .idea/modules/Machine-shop-simulator_test.iml]: these iml and xml changes should not be uploaded.  This could complicate someones gradle settings, making the project not run

File: src/applications/MachineShopSimulator.java [hunk 8]
```diff
@@ -42,40 +42,55 @@ static boolean moveToNextMachine(Job theJob, SimulationResults simulationResults
         }
     }
 
+    private static int getMachine(Job theJob) {
+        return ((Task) theJob.getTaskQ().getFrontElement()).getMachine();
+    }
```
Comment (globbycodes) [inline src/applications/MachineShopSimulator.java]: It seems like the message chain is still here. If something will get changed in this chain relationship, then this method will have to be changed. It also isn't easy to read and follow through.
We approached this problem in a different way, we found the end point method "getMachine()" and moved it to the beginning of the chain which in this case is "Job" class.

File: src/applications/MachineShopSimulator.java [hunk 9]
```diff
@@ -42,40 +42,55 @@ static boolean moveToNextMachine(Job theJob, SimulationResults simulationResults
...
-    static Job changeState(int theMachine) {// Task on theMachine has finished,
-                                            // schedule next one.
-        Job lastJob;
-        if (machine[theMachine].getActiveJob() == null) {// in idle or change-over
-                                                    // state
-            lastJob = null;
-            // wait over, ready for new job
-            if (machine[theMachine].getJobQ().isEmpty()) // no waiting job
+    static Job changeState(int theMachine) {// Task on theMachine has finished, schedule next one.
+        Machine currentMachine = machine[theMachine];
```
Comment (globbycodes) [inline src/applications/MachineShopSimulator.java:52]: Nice variable assigning, makes code look better.

**Review by globbycodes — CHANGES_REQUESTED** [review]
There a are couple things that I would've done differently.

**Comment by globbycodes** [comment]
Pull request documentation is good, I can see what issues it will address.

**Review by holma198 — CHANGES_REQUESTED** [review]
### Code Review - RJ Holman
commit 3 [7f3b3e45f79be56c7da2b1bafec0696674ac3e3b] includes the addition of `xml` and `iml` files which could mess up the project's gradle settings depending on the location of the project.  
 ##### Documentation
I think an area of improvement could definitely be the lack of details in both the issues and the pull request.  There are some very small issues which probably don’t need that much explanation, but should probably explain why the issue is being called.  More of an issue is the Epic, which has no description.  The only area of description lies in the pull request itself which is lacking in detail and specifics of what was changed.  Ideally the pull request should have everything that I need to know, however, it doesn’t.  These are more minor gripes but they should be addressed nonetheless. 
 ##### Commits
There are only 4 commits here, #4 could’ve had without the iml and xml files, and #1 could’ve been split into two separate commits, one for adding the gets, and one for replacing the arrays with the get methods
 ##### refactoring quality
I think that this has been improved, but only marginally.  It is a bit more readable, and reduces some of the reliance on the array data structures, however, it did add more code, and the pull request did not address whether ALL of the message chains had been addressed, other than that the issue has been closed.  I do not believe that the message chain issue has been completely resolved however.  There are several class files that communicate in awkward ways.  These definitely add to the message chains issue.
 #### Bottom Line
 - remove `iml` and `xml` changes from [7f3b3e45f79be56c7da2b1bafec0696674ac3e3b]
 - I'm concerned whether these commits truly close #4 or simply address it, I would suggest to say address #4 instead of closes #4

**Review by Xaivteev — CHANGES_REQUESTED** [review]
Your documentation could be better in the issues (no description), and in the pull request. A slightly longer description of what the issues were and how your changes resolved them could help.

While the original methods are now much easier to read, the message chains still exist within the new methods written. To further reduce this, methods could have been written in other classes to get the relevant information. 

For instance, instead of getMachine() in MachineShopSimulator, a method could have been written in Job to get the machine at the front of the taskQ (which makes the method make more sense here because this is where the taskQ is). 

This would give the code a similar amount of readability within the main methods:

(theJob.getNextMachine() vs getMachine(theJob))

But would reduce the message chaining in the new methods written:

(taskQ.getFrontElement().getMachine() vs theJob.getTaskQ().getFrontElement().getMachine()).

File: src/applications/MachineShopSimulator.java (hunk 8 above)
Comment (CourtneyJC) [reply src/applications/MachineShopSimulator.java]: I agree, you just moved the chain elsewhere

File: src/applications/MachineShopSimulator.java (hunk 2 above)
Comment (CourtneyJC) [reply src/applications/MachineShopSimulator.java:80]: There's still a lot of chaining going on, though it is easier to understand now.

File: src/applications/MachineShopSimulator.java [hunk 10]
```diff
@@ -51,25 +51,23 @@ private static int getMachine(Job theJob) {
...
                 getNextJob(currentMachine);
                 setReturnTime(currentMachine);
                 incrementTasks(currentMachine);
                 int t = currentMachine.getActiveJob().removeNextTask();
                 eList.setFinishTime(theMachine, timeNow + t);
             }
-        } else {// task has just finished on currentMachine
-                // schedule change-over time
+        } else {
+            // task has just finished on currentMachine schedule change-over time
```
Comment (CourtneyJC) [inline src/applications/MachineShopSimulator.java:67]: This can be confusing, throw a comma in there where the line break was before.

**Review by CourtneyJC — CHANGES_REQUESTED** [review]
Your epic is not complete. You've made progress toward getting rid of message chains, but it's not finished, so don't mark it as complete in this pull request. Also .iml and .xml files, but everyone's already said that

File: src/applications/MachineShopSimulator.java (hunk 3 above)
Comment (twarling) [reply src/applications/MachineShopSimulator.java]: I agree. This code is a lot easier to read than it was before.

**Review by twarling — CHANGES_REQUESTED** [review]
I found that the code is easier to read than before due to the addition of methods with descriptive names. However, the new methods you created still contains message chains that are hard to understand. I would recommend trying to break them up a little more.

**Review by rabex028 — CHANGES_REQUESTED** [review]
The code in MachineShopSimulator seems much easier to understand, especially in the changeState method, so that's good to see. I would have to agree with several other reviewers that I don't think the message chain issue has been resolved since the chains have merely been extracted into their own functions. I would recommend trying to eliminate them completely before performing a merge, though I think you did still make improvements to the quality of the code in some gnarly methods.

File: .gitignore [hunk 11]
```diff
@@ -44,6 +44,10 @@ cmake-build-*/
 
 # IntelliJ
 out/
+.idea/
```
Comment (leonidscott) [inline .gitignore:47]: :thumbsup:

File: src/applications/MachineShopSimulator.java (hunk 1 above)
Comment (leonidscott) [reply src/applications/MachineShopSimulator.java]: You should consider changing the variable `p`. What does it have to do with a machine? Maybe `machine` would be a better variable name.

File: src/applications/MachineShopSimulator.java [hunk 12]
```diff
@@ -42,40 +42,55 @@ static boolean moveToNextMachine(Job theJob, SimulationResults simulationResults
...
-                                            // schedule next one.
-        Job lastJob;
-        if (machine[theMachine].getActiveJob() == null) {// in idle or change-over
-                                                    // state
-            lastJob = null;
-            // wait over, ready for new job
-            if (machine[theMachine].getJobQ().isEmpty()) // no waiting job
+    static Job changeState(int theMachine) {// Task on theMachine has finished, schedule next one.
+        Machine currentMachine = machine[theMachine];
+        Job lastJob = null;
```
Comment (leonidscott) [inline src/applications/MachineShopSimulator.java:53]: Why do you need to set this to `null`?

File: src/applications/MachineShopSimulator.java [hunk 13]
```diff
@@ -42,40 +42,55 @@ static boolean moveToNextMachine(Job theJob, SimulationResults simulationResults
...
+        if (currentMachine.getActiveJob() == null) {
+            // in idle or change-over state wait over, ready for new job
```
Comment (leonidscott) [inline src/applications/MachineShopSimulator.java:55]: Nice job consolidating comments

File: src/applications/MachineShopSimulator.java (hunk 2 above)
Comment (leonidscott) [reply src/applications/MachineShopSimulator.java:80]: I would argue that the bulk of the logic in this method should be exported to another class. You should decide which object should take the responsibility of doing this work, and add that functionality there.

**Review by leonidscott — CHANGES_REQUESTED** [review]
Your PR does not make things worse, but just like @holma198 said, it doesn't make thing too much better. You can clearly see that removing the message chain out from the function does increasing readability a lot, but that functionality should really be one of the objects (in most cases, either `Job`, or `Machine`).  

It was nice to see a lot of issues linked together by an *epic*, however, I had no idea what most of those issues means because there is little documentation. However, I did like that they were small, and able to be conquered in only a few commits. I don't necessarily think that you had too few commits considering how small the impact of each commit was.  

Before accepting, I would like to see:  
* A description of each issue and epic.  
* The destructuring of message chains in objects rather than local functions. You will most likely see that a lot of the steps can either be split up further, or that many of them are just flat our unnecessary.

**Review by twarling — APPROVED** [review]
